pkg/fqdn/dnsproxy/udp.go

```diff
@@ -0,0 +1,250 @@
+func transparentSetsockopt(fd int) error {
+	err6 := unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_TRANSPARENT, 1)
+	if err6 == nil {
+		err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_RECVORIGDSTADDR, 1)
+	}
+	err4 := unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_TRANSPARENT, 1)
+	if err4 == nil {
+		err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_RECVORIGDSTADDR, 1)
+	}
+	// Only error out if both IPv4 and IPv6 fail
+	if err4 != nil && err6 != nil {
+		return err4
+	}
+	return nil
+}
+
+func listenConfig(mark int) *net.ListenConfig {
+	return &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			var opErr error
+			err := c.Control(func(fd uintptr) {
+				opErr = transparentSetsockopt(int(fd))
+				if opErr == nil && mark != 0 {
+					opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, mark)
+				}
+				if opErr == nil {
+					opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
+				}
+			})
+			if err != nil {
+				return err
+			}
+
+			return opErr
+		}}
+}
+
+func bindUDP(addr string) *net.IPConn {
+	// Mark outgoing packets as proxy egress return traffic (0x0b00)
+	conn, err := listenConfig(0xb00).ListenPacket(context.Background(), "ip:udp", addr)
+	if err != nil {
+		log.Errorf("bindUDP failed for address %s: %s", addr, err)
+		return nil
+	}
+	return conn.(*net.IPConn)
+}
+
+func (f *sessionUDPFactory) CreateSessionUDP(msgSize int) dns.SessionUDP {
+	return &sessionUDP{
+		m:   make([]byte, msgSize),
+		oob: make([]byte, udpOOBSize),
+	}
+}
+
+var udpOnce sync.Once
+
+func (s *sessionUDPFactory) InitConn(conn *net.UDPConn) error {
+	// Set up the raw socket for sending responses.
+	// - Must use a raw UDP socket for sending responses so that we can send
+	//   from a specific port without binding to it.
+	// - The raw UDP socket must be bound to a specific IP address to prevent
+	//   it receiving ALL UDP packets on the host.
+	// - We use oob data to override the source IP address when sending
+	// - Must use separate sockets for IPv4/IPv6, as sending to a v6-mapped
+	//   v4 address from a socket bound to "::1" does not work due to kernel
+	//   checking that a route exists from the source address before
+	//   the source address is replaced with the (transparently) changed one
+	udpOnce.Do(func() {
+		rawconn4 = bindUDP("127.0.0.1") // raw socket for sending IPv4
+		rawconn6 = bindUDP("::1")       // raw socket for sending IPv6
+	})
+	if rawconn4 == nil || rawconn6 == nil {
+		return fmt.Errorf("Unable to open raw UDP sockets for DNS Proxy")
+	}
+	return nil
+}
+
+// Clear re-initializes sessionUDP to the same state it was when new.
+// Returns the interface for convenience.
+// This is required for pooling of dns.SessionUDP objects.
+func (s *sessionUDP) Clear() dns.SessionUDP {
+	s.conn = nil
+	s.raddr = nil
+	s.laddr = nil
+	s.m = s.m[:cap(s.m)]
+	s.oob = s.oob[:cap(s.oob)]
```

Comment:
I don't understand this.

1) how can you guarantee the slice hasn't increased its capacity?
2) I think this needs to be at least `make([]byte, udpOOBSize, cap(s.oob))`. Similar above `make([]byte, len(s.m) ,msgSize)` so we are sure the slice is really "clean" up

Reply:
The above changes the length of the slice to be the same as it's capacity. syscall.Recvmsg() gets the oob slice and only uses it upto the length of it.

Initialization `make([]byte, udpOOBSize)` creates a new slice with the given capacity, initializes the elements to zero values and returns with `len() == udpOOBSize`. As such the slice is ready to be used with `syscall.Recvmsg()`. After the UDP IO we re-slice the data to the actual number of bytes returned by the syscall, so then `len() <= cap()`. Syscall never increases the capacity.

In `Clear()` (quoted above), we re-slice `oob` (respecitively, `m`) so that `len()` is again the same as the full capacity (`cap()`). The contents need not be re-initialized, these are read buffers.

So:
1. Yes, and even if it did, it should not matter. 
2. It's better to avoid reallocating the memory.